main.py

```diff
@@ -13,21 +14,51 @@ def index():
 
 @app.route("/search", methods=["GET"])
 def search():
-    app.logger.info(f"search(): GET {request.full_path}.")
-    keyword = request.args.get("query", None)
-    if keyword:
-        context_dict = {
-            "subtitle": f"Search Result for {keyword}",
-            "keyword": keyword,
-            "message": None
-        }
+    app.logger.info(f"search(): GET /{request.full_path}")
+    context_dict = {
+        "subtitle": "Search results for {{ keyword }}",
+        "keyword": request.args.get("query", "")
+    }
+    if session.get("product_list", ""):
+        context_dict["product_list"] = session["product_list"]
+        context_dict["item_hits"] = len(session["product_list"])
         return render_template("search.html", **context_dict)
+    if context_dict["keyword"]:
+        try:
+            session["product_list"] = amazon_api_client.search_products(keywords=context_dict["keyword"])
+            context_dict["product_list"] = session["product_list"]
+            context_dict["item_hits"] = len(session["product_list"])
+
+        except AmazonException as ae:
+            app.logger.error(f"{ae}")
+            raise ae
+        return render_template("search.html", **context_dict)
+
     else:
-        context_dict = {
-            "message": "Enter any keywords."
-        }
+        context_dict["message"] = "Enter any keywords."
         return render_template("index.html", **context_dict)
 
 
+@ app.route("/registration", methods=["GET", "POST"])
+def registration():
+    if request.method == "GET":
+        app.logger.info(f"registration: GET /{request.full_path}")
+        return render_template("index.html", message="Enter any keywords.")
+    else:
+        app.logger.info(f"registration: POST /{request.full_path}")
```

Comment:
elseがいらなくって、debugレベルでformの内容をログ出力しても良さそう？
```suggestion
    app.logger.info(f"registration: POST {request.full_path}")
    app.logger.debug(f"{request.form=}")
```